### PR TITLE
Goal: Introduce new flags to make creating apps without state easier

### PR DESCRIPTION
**Pull request by jasonpaulos**
## Summary

With the recent introduction of boxes, now more than ever it makes sense for some apps to not use global/local state.

This PR introduces some new convenience flags for `goal app create` and `goal method --create` which makes it easier to create an app without global, local, or neither states. They are:
* `--no-global-state`: shorthand for `--global-ints=0` and `--global-byteslices=0`
* `--no-local-state`: shorthand for `--local-ints=0` and `--local-byteslices=0`
* `--no-state`: shorthand for `--no-local-state` and `--no-global-state`

For example, this simplifies usage from `goal app create --creator ${ACCOUNT} --approval-prog ${TEALDIR}/boxes.teal --clear-prog ${TEALDIR}/clear.teal --global-byteslices 0 --global-ints 0 --local-byteslices 0 --local-ints 0` to `goal app create --creator ${ACCOUNT} --approval-prog ${TEALDIR}/boxes.teal --clear-prog ${TEALDIR}/clear.teal --no-state
`

## Test Plan

New test added in `test/scripts/e2e_subs/goal-app-create-state.sh`, and I modified some existing e2e tests to take advantage of the new flags where appropriate.

### CODE REVIEW

**Review by jannotti — APPROVED** [review]
I've hated the requirement on these switches for these for some time, so I like the idea.

I assume you considered simply not making them required, thus defaulting to 0?  That's what I would have pushed for if we were starting from scratch, but maybe you don't feel we should make that change?

The advantage of that is that it lets you conveniently specify only what you care about, rather than needing to specify that the other stuff is 0 as soon as you want to set any.

I'm putting in approval, because this is fine with me if you want to go this way, but I think I'd prefer making them optional instead, even now. After all, since they were required, it can't break any existing scripts/code.

File: cmd/goal/application.go [hunk 1]
```diff
@@ -396,6 +402,77 @@ func mustParseOnCompletion(ocString string) (oc transactions.OnCompletion) {
...
+		noGlobalStateSet:          cmd.Flags().Changed("no-global-state"),
+		noLocalStateSet:           cmd.Flags().Changed("no-local-state"),
+		globalSchemaUintsSet:      cmd.Flags().Changed("global-ints"),
+		globalSchemaByteSlicesSet: cmd.Flags().Changed("global-byteslices"),
+		localSchemaUintsSet:       cmd.Flags().Changed("local-ints"),
+		localSchemaByteSlicesSet:  cmd.Flags().Changed("local-byteslices"),
+	}
+}
+
+func getStateSchema(cmd *cobra.Command) basics.StateSchemas {
```
Comment (bbroder-algo) [inline cmd/goal/application.go:437]: checks look good

File: test/scripts/e2e_subs/goal-app-create-state.sh [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+gcmd="goal -w ${WALLET}"
+
+ACCOUNT=$(${gcmd} account list|awk '{ print $3 }')
+
+printf '#pragma version 2\nint 1' > "${TEMPDIR}/simple.teal"
+
+# Check goal flags --no-state, --no-local-state, --no-global-state
+
```
Comment (bbroder-algo) [inline test/scripts/e2e_subs/goal-app-create-state.sh:23]: maybe a few extra words here about the methodology

**Comment by bbroder-algo** [comment]
This is definitely better in every way.  But maybe not better than defaulting them to zero?

**Comment by jasonpaulos** [comment]
After more thought I think I agree it would be better to allow the existing arguments to be optional. If I have time in the next few days, I'll modify this PR to do that instead.

**Comment by bbroder-algo** [comment]
how u doin

**Comment by jasonpaulos** [comment]
Closing this in favor of a future PR which makes the existing arguments optional

**Pull request closed without merging**